### PR TITLE
Adding support for Sse2.StoreScalar for int/uint

### LINKED ISSUE #31179: API Proposal: Sse2.StoreScalar(int*)
## Background

`System.Runtime.Intrinsics.X86.Sse2` exposes load and store methods accepting address arguments for most supported types, but methods to store to an `int*` are missing.

The load versions of the methods are present:

```C#
/// <summary>
/// __m128i _mm_loadl_epi32 (__m128i const* mem_addr)
/// MOVD xmm, reg/m32
/// The above native signature does not exist. We provide this additional overload for completeness.
/// </summary>
public static unsafe Vector128<int> LoadScalarVector128(int* address) => LoadScalarVector128(address);
/// <summary>
/// __m128i _mm_loadl_epi32 (__m128i const* mem_addr)
/// MOVD xmm, reg/m32
/// The above native signature does not exist. We provide this additional overload for completeness.
/// </summary>
public static unsafe Vector128<uint> LoadScalarVector128(uint* address) => LoadScalarVector128(address);
```

As noted in the comments, `_mm_loadl_epi32` doesn't exist in the Intel intrinsics guide.  However, there is a matching signature under `_mm_loadu_si32`

```
__m128i _mm_loadu_si32 (void const* mem_addr)
#include <immintrin.h>
Instruction: movd xmm, m32
CPUID Flags: SSE2
```

Its counterpart for storing a scalar int32 also exists:

```
void _mm_storeu_si32 (void* mem_addr, __m128i a)
#include <immintrin.h>
Instruction: movd m32, xmm
CPUID Flags: SSE2
```

However the `StoreScalar` pair that would correspond to that intrinsic is missing.

## Rationale

`StoreScalar` is implemented with address arguments for `float`, `double` and `long`, but not for `int`.

`StoreScalar` methods for `int` would mirror the existing `LoadScalarVector128(int *)` methods and would complete the set of supported types for stores.

## Poposed API

```C#
public abstract class Sse2 : Sse
{
     /// <summary>
     /// void _mm_storeu_si32 (void* mem_addr, __m128i a)
     ///   MOVD m32, xmm
     /// </summary>
    public static unsafe void StoreScalar(int* address, Vector128<int> source);
     /// <summary>
     /// void _mm_storeu_si32 (void* mem_addr, __m128i a)
     ///   MOVD m32, xmm
     /// </summary>
    public static unsafe void StoreScalar(uint* address, Vector128<uint> source);
}
```

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/runtime/issues/31179.

CC. @CarolEidt, @echesakovMSFT

### CODE REVIEW

**Comment by tannergooding** [comment]
The 3rd commit regenerates the x86 hardware intrinsic tests from the template and so reviewing the template changes in the second commit should be sufficient (rather than reviewing every generated test).

**Comment by saucecontrol** [comment]
Would be good to update the XML comments for the `movd` `LoadScalarVector128` overloads to the correct native intrinsic mapping (`_mm_loadu_si32`) and remove the note about said intrinsic not existing since this PR is fixing the load/store parity issue.

https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Sse2.cs#L780-L791

**Comment by tannergooding** [comment]
> Would be good to update the XML comments for the movd LoadScalarVector128 overloads

Fixed.

**Comment by EgorBo** [comment]
heh, once the managed API lands it will be also supported by mono-llvm without runtime changes since the implementation is type-agnostic: https://github.com/dotnet/runtime/blob/25fdaa850f492a9b4144670cac3522bd5b57cd6f/src/mono/mono/mini/mini-llvm.c#L7589-L7594

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Sse2.PlatformNotSupported.cs [hunk 1]
```diff
@@ -774,15 +774,13 @@ public new abstract class X64 : Sse.X64
         public static unsafe Vector128<double> LoadLow(Vector128<double> upper, double* address) { throw new PlatformNotSupportedException(); }
 
         /// <summary>
-        /// __m128i _mm_loadl_epi32 (__m128i const* mem_addr)
+        /// __m128i _mm_loadu_si32 (void const* mem_addr)
         ///   MOVD xmm, reg/m64
```
Comment (saucecontrol) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Sse2.PlatformNotSupported.cs]: While you're in there, these should also be `m32` 😄

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Sse2.PlatformNotSupported.cs]: Fixed. Looks like this one was just in the PNSE version of the file.

**Comment by tannergooding** [comment]
This should be ready for review/merge.

**Review by echesakov — APPROVED** [review]

**Pull request merged**